src/MongoDB.Driver/PipelineStageDefinitionBuilder.cs

```diff
@@ -927,6 +927,108 @@ public static class PipelineStageDefinitionBuilder
             return Match(new ExpressionFilterDefinition<TInput>(filter));
         }
 
+        /// <summary>
+        /// Creates a $merge stage.
+        /// </summary>
+        /// <typeparam name="TInput">The type of the input documents.</typeparam>
+        /// <typeparam name="TOutput">The type of the output documents.</typeparam>
+        /// <param name="outputCollection">The output collection.</param>
+        /// <param name="mergeOptions">The merge options.</param>
+        /// <returns>The stage.</returns>
+        public static PipelineStageDefinition<TInput, TOutput> Merge<TInput, TOutput>(
+            IMongoCollection<TOutput> outputCollection,
+            MergeStageOptions<TOutput> mergeOptions)
+        {
+            Ensure.IsNotNull(outputCollection, nameof(outputCollection));
+            Ensure.IsNotNull(mergeOptions, nameof(mergeOptions));
+
+            if (mergeOptions.LetVariables != null && mergeOptions.WhenMatched != MergeStageWhenMatched.Pipeline)
+            {
+                throw new ArgumentException("LetVariables can only be set when WhenMatched == Pipeline.");
+            }
+
+            if (mergeOptions.WhenMatchedPipeline == null)
```

Comment:
The checks on lines 950-963 are different than the one in lines 945-948.

To the server `whenMatched` is a single polymorphic field.

The the C# driver it's an enum paired with an optional pipeline that is only valid when the enum value is `Pipeline`.